src/java/com/puppetlabs/http/client/impl/JavaClient.java

```diff
@@ -307,35 +309,34 @@ public void cancelled() {
         client.execute(HttpAsyncMethods.create(request), consumer, streamingCompleteCallback);
     }
 
-    public static Promise<Response> requestWithClient(final RequestOptions requestOptions,
-                                                      final HttpMethod method,
-                                                      final IResponseCallback callback,
-                                                      final CloseableHttpAsyncClient client) {
+    public static void requestWithClient(final RequestOptions requestOptions,
+                                         final HttpMethod method,
+                                         final IResponseCallback callback,
+                                         final CloseableHttpAsyncClient client,
+                                         final ResponseDeliveryDelegate responseDeliveryDelegate) {
 
-        CoercedRequestOptions coercedRequestOptions = coerceRequestOptions(requestOptions, method);
+        final CoercedRequestOptions coercedRequestOptions = coerceRequestOptions(requestOptions, method);
 
-        HttpRequestBase request = constructRequest(coercedRequestOptions.getMethod(),
+        final HttpRequestBase request = constructRequest(coercedRequestOptions.getMethod(),
                 coercedRequestOptions.getUri(), coercedRequestOptions.getBody());
         request.setHeaders(coercedRequestOptions.getHeaders());
 
         final HttpContext httpContext = HttpClientContext.create();
 
-        final Promise<Response> promise = new Promise<>();
-
-        FutureCallback futureCallback = new FutureCallback<HttpResponse>() {
+        final FutureCallback<HttpResponse> futureCallback = new FutureCallback<HttpResponse>() {
             @Override
             public void completed(HttpResponse httpResponse) {
-                completeResponse(promise, requestOptions, callback, httpResponse, httpContext);
+                completeResponse(responseDeliveryDelegate, requestOptions, callback, httpResponse, httpContext);
             }
 
             @Override
             public void failed(Exception e) {
-                deliverResponse(requestOptions, new Response(requestOptions, e), callback, promise);
+                responseDeliveryDelegate.deliverResponse(requestOptions, e);
```

Comment:
If a callback was provided for the request, will it still be invoked if the request fails?